lib/std/Build.zig

```diff
@@ -1512,6 +1505,248 @@ pub const Module = struct {
     /// file of directory of files which constitute the module.
     source_file: FileSource,
     dependencies: std.StringArrayHashMap(*Module),
+    include_dirs: std.ArrayList(CompileStep.IncludeDir),
+    lib_paths: std.ArrayList([]const u8),
+    rpaths: std.ArrayList([]const u8),
+    framework_dirs: std.ArrayList([]const u8),
+    system_libs: std.ArrayList(CompileStep.SystemLib),
+    libs: std.ArrayList(*CompileStep),
+    config_headers: std.ArrayList(*ConfigHeaderStep),
+    installed_headers: std.ArrayList(InstalledHeader),
+    frameworks: std.StringArrayHashMap(CompileStep.FrameworkLinkInfo),
+
+    const InstalledHeader = union(enum) {
+        header: struct {
+            source_path: []const u8,
+            dest_rel_path: []const u8,
+        },
+        header_dir_step: InstallDirStep.Options,
+        config_header: struct {
+            step: *ConfigHeaderStep,
+            options: CompileStep.InstallConfigHeaderOptions,
+        },
+        lib_step: *CompileStep,
+    };
+
+    pub fn init(arena: Allocator, builder: *Builder, source_file: FileSource, dependencies: []const ModuleDependency) Module {
+        return .{
+            .builder = builder,
+            .source_file = source_file,
+            .dependencies = moduleDependenciesToArrayHashMap(arena, dependencies),
+            .include_dirs = std.ArrayList(CompileStep.IncludeDir).init(arena),
+            .lib_paths = std.ArrayList([]const u8).init(arena),
+            .rpaths = std.ArrayList([]const u8).init(arena),
+            .framework_dirs = std.ArrayList([]const u8).init(arena),
+            .system_libs = std.ArrayList(CompileStep.SystemLib).init(arena),
+            .libs = std.ArrayList(*CompileStep).init(arena),
+            .config_headers = std.ArrayList(*ConfigHeaderStep).init(arena),
+            .installed_headers = std.ArrayList(InstalledHeader).init(arena),
+            .frameworks = std.StringArrayHashMap(CompileStep.FrameworkLinkInfo).init(arena),
+        };
+    }
+
+    fn moduleDependenciesToArrayHashMap(arena: Allocator, deps: []const ModuleDependency) std.StringArrayHashMap(*Module) {
+        var result = std.StringArrayHashMap(*Module).init(arena);
+        for (deps) |dep| {
+            result.put(dep.name, dep.module) catch @panic("OOM");
+        }
+        return result;
+    }
+
+    pub fn deinit(m: *Module) void {
+        m.dependencies.deinit();
+        m.include_dirs.deinit();
+        m.lib_paths.deinit();
+        m.rpaths.deinit();
+        m.framework_dirs.deinit();
+        m.system_libs.deinit();
+        m.libs.deinit();
+        m.config_headers.deinit();
+        m.installed_headers.deinit();
+        m.frameworks.deinit();
+        m.* = undefined;
+    }
+
+    pub fn addIncludePath(m: *Module, path: []const u8) void {
+        m.include_dirs.append(
+            CompileStep.IncludeDir{ .raw_path = m.builder.pathFromRoot(path) },
```

Comment:
Calls to `pathFromRoot` should not be happening in the configure phase. This is premature.

(many similar instances follow)